src/Featurizers/Components/NormUpdaters.h

```diff
@@ -1,298 +1,373 @@
-    }
-    return std::sqrt(_l2_norm);
-}
-// ----------------------------------------------------------------------
-// |
-// |  Updaters::MaxNormUpdater
-// |
-// ----------------------------------------------------------------------
-template <typename T>
-Updaters::MaxNormUpdater<T>::MaxNormUpdater() :
-    _max_norm(0),
-    _update_flag(false) {
-}
-
-template <typename T>
-void Updaters::MaxNormUpdater<T>::update(T input) {
-    if (!_update_flag) {
-        _max_norm = static_cast<typename TypeSelector::MaxNormTypeSelector<T>::type>(std::abs(input));
-        _update_flag = true;
-    } else {
-        _max_norm = std::max(static_cast<typename TypeSelector::MaxNormTypeSelector<T>::type>(std::abs(input)), _max_norm);
-    }
-}
-
-
-template <typename T>
-typename TypeSelector::MaxNormTypeSelector<T>::type Updaters::MaxNormUpdater<T>::commit(void) {
-    assert(_max_norm >= 0);
-    if (!_update_flag) {
-        throw std::runtime_error("update is not called before max is committed!");
-    }
-    return _max_norm;
-}
-
-#if (defined __clang__)
-#   pragma clang diagnostic pop
-#endif
-
-} // namespace Components
-} // namespace Featurizers
-} // namespace Featurizer
-} // namespace Microsoft
+// ----------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License
+// ----------------------------------------------------------------------
+#pragma once
+
+#include <limits>
+#include "../../Traits.h"
+
+namespace Microsoft {
+namespace Featurizer {
+namespace Featurizers {
+namespace Components {
+
+namespace TypeSelector {
+
+// max norm type selector is created for selecting different types for max norm based on input type
+// when input type is integer, uint64 is chosen for performance
+// when input type is numeric, long double is chosen
+template <typename T>
+struct MaxNormTypeSelector {
+    using type = std::uint64_t;
+};
+template <>
+struct MaxNormTypeSelector<std::float_t> {
+    using type = long double;
+};
+
+template <>
+struct MaxNormTypeSelector<std::double_t> {
+    using type = long double;
+};
+
+} // namespace TypeSelector
+
+#if (defined __clang__)
+#   pragma clang diagnostic push
+#   pragma clang diagnostic ignored "-Wfloat-equal"
+#   pragma clang diagnostic ignored "-Wdouble-promotion"
+#endif
+
+namespace Updaters{
+
+/////////////////////////////////////////////////////////////////////////
+///  \class         L1NormUpdater
+///  \brief         L1NormUpdater updates l1 norm
+///                 l1 norm is the sum of elements absolute value
+///
+template <typename T>
+class L1NormUpdater {
+public:
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Types
+    // |
+    // ----------------------------------------------------------------------
+    using InputType                         = T;
```

Comment:
>using InputType                         = T; [](start = 4, length = 44)

InputType isn't a concept that directly applies to this class - T should be sufficient.